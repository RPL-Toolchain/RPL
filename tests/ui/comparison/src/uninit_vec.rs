#[cfg_attr(test, test)]
fn base_case() {
    // with_capacity() -> set_len() should be detected
    let mut vec: Vec<u8> = Vec::with_capacity(1000);
    //~^ uninit_vec

    unsafe {
        vec.set_len(200);
    }

    dbg!(vec[0]);
}

#[cfg_attr(test, test)]
fn cross_function_with_capacity() {
    fn with_capacity<T>(capacity: usize) -> Vec<T> {
        Vec::with_capacity(capacity)
    }
    // with_capacity() -> set_len() should be detected
    let mut vec: Vec<u8> = with_capacity(1000);
    //~^ uninit_vec

    unsafe {
        vec.set_len(200);
    }

    dbg!(vec[0]);
}

#[cfg_attr(test, test)]
fn cross_function_set_len() {
    unsafe fn set_len<T>(v: &mut Vec<T>, len: usize) {
        v.set_len(len);
    }
    // with_capacity() -> set_len() should be detected
    let mut vec: Vec<u8> = Vec::with_capacity(1000);
    //~^ uninit_vec

    unsafe {
        set_len(&mut vec, 200);
    }

    dbg!(vec[0]);
}

fn cross_function_both() {
    fn with_capacity<T>(capacity: usize) -> Vec<T> {
        Vec::with_capacity(capacity)
    }
    unsafe fn set_len<T>(v: &mut Vec<T>, len: usize) {
        v.set_len(len);
    }
    // with_capacity() -> set_len() should be detected
    let mut vec: Vec<u8> = with_capacity(1000);
    //~^ uninit_vec

    unsafe {
        set_len(&mut vec, 200);
    }

    dbg!(vec[0]);
}

pub(crate) fn main() {
    base_case();
    cross_function_with_capacity();
    cross_function_set_len();
    cross_function_both();
}
