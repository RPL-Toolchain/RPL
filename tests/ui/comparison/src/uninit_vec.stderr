error[E0133]: call to unsafe function `std::vec::Vec::<T, A>::set_len` is unsafe and requires unsafe block
  --> tests/ui/comparison/src/uninit_vec.rs:33:9
   |
LL |         v.set_len(len);
   |         ^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2024/unsafe-op-in-unsafe-fn.html>
   = note: consult the function's documentation for information on how to avoid undefined behavior
note: an unsafe function restricts its caller, but its body is safe by default
  --> tests/ui/comparison/src/uninit_vec.rs:32:5
   |
LL |     unsafe fn set_len<T>(v: &mut Vec<T>, len: usize) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: `-D unsafe-op-in-unsafe-fn` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(unsafe_op_in_unsafe_fn)]`

error[E0133]: call to unsafe function `std::vec::Vec::<T, A>::set_len` is unsafe and requires unsafe block
  --> tests/ui/comparison/src/uninit_vec.rs:51:9
   |
LL |         v.set_len(len);
   |         ^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2024/unsafe-op-in-unsafe-fn.html>
   = note: consult the function's documentation for information on how to avoid undefined behavior
note: an unsafe function restricts its caller, but its body is safe by default
  --> tests/ui/comparison/src/uninit_vec.rs:50:5
   |
LL |     unsafe fn set_len<T>(v: &mut Vec<T>, len: usize) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0133`.
